pkg/module/oci/client.go

```diff
@@ -0,0 +1,216 @@
+package oci
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/opencontainers/go-digest"
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const (
+	DefaultTimeout = 10 * time.Second
+)
+
+type Client interface {
+	// GetManifest returns the ocispec Manifest for a reference
+	GetManifest(ctx context.Context, ref string) (*ocispecv1.Manifest, error)
+
+	// PushManifest uploads the given manifest with all its layers to the given reference in the registry configured in the client.
+	PushManifest(ctx context.Context, ref string, manifest *ocispecv1.Manifest) error
+
+	// Cache exposes the client's cache where all intermediate manifests and data are stored.
+	Cache() Cache
+}
+
+type client struct {
+	cache    Cache
+	registry string
+	// user to authenticate when calling the registry configured in the client
+	user string
+	// secret can be either a password (if user provided) or a long lived token.
+	secret string
+	// timeout for all network calls the client will perform
+	timeout time.Duration
+	// if true, the client will make all calls with http instead of https
+	insecure bool
+}
+
+type Options struct {
+	Registry string
+	// (Optional) user to authenticate when calling the registry configured in the client
+	User string
+	// (Optional) secret can be either a password (if user provided) or a long lived token.
+	Secret string
+	// timeout for all network calls the client will perform
+	Timeout time.Duration
+	// if true, the client will make all calls with http instead of https
+	Insecure bool
+}
+
+func NewClient(o *Options) (Client, error) {
+	c := &client{
+		registry: o.Registry,
+		user:     o.User,
+		secret:   o.Secret,
+		timeout:  o.Timeout,
+		insecure: o.Insecure,
+	}
+
+	if c.timeout == 0 {
+		c.timeout = DefaultTimeout
+	}
+
+	c.cache = NewInMemoryCache()
+
+	return c, nil
+}
+
+func (c *client) Cache() Cache {
+	return c.cache
+}
+
+func (c *client) GetManifest(ctx context.Context, ref string) (*ocispecv1.Manifest, error) {
+	// TODO
+	return nil, nil
+}
+
+func (c *client) PushManifest(ctx context.Context, ref string, manifest *ocispecv1.Manifest) error {
+	manifestBytes, err := json.Marshal(manifest)
+	if err != nil {
+		return fmt.Errorf("unable to marshal manifest: %w", err)
+	}
+
+	desc := ocispecv1.Descriptor{
+		MediaType:   ocispecv1.MediaTypeImageManifest,
+		Digest:      digest.FromBytes(manifestBytes),
+		Size:        int64(len(manifestBytes)),
+		Annotations: manifest.Annotations,
+	}
+
+	resolver, err := c.resolver()
+	if err != nil {
+		return err
+	}
+
+	pusher, err := resolver.Pusher(ctx, ref)
+	if err != nil {
+		return err
+	}
+
+	if isSingleArchImage(desc.MediaType) {
```

Comment:
if its not single archive, wouldn't it error out after the rest?

Reply:
I have not tested that case TBH, but if it is not single arch it should just push the manifest as is. I don't see a reason why that would error.